changed impl computed_assign st shapes of different types are supported

DerThorsten opened the pull request:
With this pr the following is compiling:
```C++
pyarray<int>    a1  { 0,  1,  2};
xt::xarray<int> a2  { 0,  2,  4};
xt::noalias(a1) += a2;
```

```C++
xt::xtensor_fixed<int, xt::xshape<2>> a1 = {2,3};
xt::xarray<int> a2{ 0,  2,  4};
xt::noalias(a1) += a2;
```

JohanMabille reviewed (changes requested):
I think the mechanism for `fixed` containers should actually be in a dedicated function, that would check that `shape == de1.shape()` and throw otherwise

Indeed, broadcasting in the following expression `a += b` where `a` is a fixed container does not really make sense. This would simplify this PR, we can push the fix for fixed containers in a dedicated PR.

DerThorsten commented:
> Indeed, broadcasting in the following expression a += b where a is a fixed container does not really make sense. This would simplify this PR, we can push the fix for fixed containers in a dedicated PR.

Why would xtensor_fixed have a different behavior  than lets say an xtensor? When doing a+=b , the shape of a does not change in any case right? So we should have the same semantics for  xtensor_fixed, or do I miss something.

So Just that we are on the same side:
lets say a and b have shape the shapes:
a = (2,2)
b = (1,2) 
so here , when doing a+=b,  broadcasting makes sense (no mather if with xtensor_fixed or not)

But with:
a = (1,2)
b = (2,2) 
when doing a+=b  broadcasting does not make sense in any case (this is not allowed right? )

DerThorsten commented:
Your right, `a += b` with broadcasting does not make sense for a fixed container when `a+=b` can change the shape of a.
I was a bit surprised that an expression like `a+=b` indeed can change the shape of an `a`.

 I was still thinking in the numpy way:
In python the following would lead to a runtime error, while with xtensor `a` would just have the shape `(2,2)` after `a+=b`.
``` python
import numpy
a = numpy.zeros([1, 2])
b = numpy.zeros([2, 1])
a += b
```


Anyhow, It makes sense to handle `fixed` in a dedicated function

DerThorsten commented:
I removed the handling of the fixed function. We should take take of the fixed function in a different PR.
To trigger the compile error I hand to used a custom xarray_container:
```c++
  // xarray like with custom shape
  using my_xarray = xt::xarray_container<
      std::vector<int>,
      xt::layout_type::row_major,
      my_vector<std::size_t>
  >;
  
  auto a = my_xarray::from_shape({1,3});
  auto b = xt::xtensor<int,2>::from_shape({2,3});
  xt::noalias(a) += b;
```
Unfortunately everything still compiles when `my_vector<std::size_t>`  just derives from `std::vector<std::size_t>`.
So I wrote a non-derving my_vector with the same API as std::vector. With that the former implementation of `computed_assign` did not compile. With this implementation it does.
In my initial draft I forgot the comperator in  `lexicographical_compare` to have `>` instead of `<`.
Therefore I added also an test which is sensitive to such a confusion of `>` and `<`

DerThorsten commented on include/xtensor/xassign.hpp:
```diff
@@ -413,23 +413,60 @@ namespace xt
+        template<class SHAPE>
+        struct select_shape{
+            using type = SHAPE;
+        };
```
Ok this is now anyhow removed since we decided to do the fixed container fixes in a different pr

JohanMabille commented:
Indeed, with public inheritance from `std::vector` it can find a matching overload of `operator>`. By the way, you sould never inherit publicly from `std::vector` or any class with value semantics. You can use private inheritance with using declaration to avoid typing a lot of boilerplate:

```cpp
template <class T, class A>
class my_vector : private std::vector<T, A>
{
public:

    using base_type = std::vector<T, A>;
    using value_type = typename base_type::value_type;
    // etc
    // "imports" all constructors from std::vector. This way you can write things like
    // my_vector<int> ma = {1, 2, 3}; 
    using base_type::base_type;

    using base_type::size;
    using base_type::empty;
    using base_type::begin;
    using base_type::end;
    using base_type::cbegin;
    using base_type::cend;
    using base_type::operator[];
    // etc
};
```
